### PR TITLE
agent: add disable_keep_alives configurable

### PR DESCRIPTION
**Pull request by jasonodonnell**
Follow up for https://github.com/hashicorp/vault/pull/15986.

This PR adds a new configurable, `disable_keep_alives`, which takes a string list to disable keep alives for various features in Vault Agent (`auto-auth`, `caching` and `templating`).

A small example can be found here:
```hcl
 "auto_auth" = {
  "method" = {
    "config" = {
      "role" = "app"
    }
    "type" = "kubernetes"
  }

  "sink" = {
    "config" = {
      "path" = "/home/vault/.token"
    }
    "type" = "file"
  }
}

"disable_idle_connections" = ["auto-auth", "caching", "templating"]
"disable_keep_alives" = ["auto-auth", "caching", "templating"]
"exit_after_auth" = false
"pid_file" = "/home/vault/.pid"

"template" = {
  "contents" = "{{ with secret \"secret/hashiconf\" }}{{ .Data | toJSONPretty }}{{ end }}"
  "destination" = "/vault/secrets/kv-secret"
  "error_on_missing_key" = true
}

"vault" = {
  "address" = "https://vault.vault.svc.cluster.local:8200"
}
```

### CODE REVIEW

File: api/client.go [hunk 1]
```diff
@@ -756,6 +756,24 @@ func (c *Client) MaxIdleConnections() int {
...
+	defer c.config.modifyLock.Unlock()
+
+	c.config.HttpClient.Transport.(*http.Transport).DisableKeepAlives = disable
+}
+
+func (c *Client) DisableKeepAlives() bool {
+	c.modifyLock.RLock()
+	defer c.modifyLock.RUnlock()
+	c.config.modifyLock.Lock()
+	defer c.config.modifyLock.Unlock()
```
Comment (calvn) [inline api/client.go:772]: This could be an `RLock`

File: command/agent.go [hunk 2]
```diff
@@ -507,10 +511,14 @@ func (c *AgentCommand) Run(args []string) int {
 			return 1
 		}
 
-		if config.DisableIdleConnsAutoAuth {
+		if config.DisableIdleConnsCaching {
```
Comment (swenson) [inline command/agent.go:514]: Why did this change for `proxyClient`, but not for the other clients?

Comment (jasonodonnell) [reply command/agent.go:514]: Was actually a bug I found, this should have been the proxy client.

Comment (swenson) [reply command/agent.go:514]: :+1:

**Review by swenson — APPROVED** [review]
LGTM; would be good to change that lock to an RLock though

**Review by calvn — APPROVED** [review]
PR looks good aside from the locking comment that I left!

**Pull request merged**